### PR TITLE
broadcasting user-defined functions wip

### LINKED ISSUE #101: grad doesn't work on trivial identity
I am on the official julia 1.0.1 release with all packages up-to-date.
The following returns nothing,
```julia
using AutoGrad
y = Param(1.)
f(x) = x
d = @diff f(y)
grad(d, y) # returns nothing
```
while this works:
```julia
f(x) = 1x
d = @diff f(y)
grad(d, y) # returns 1.
```

### PR DESCRIPTION
**Pull request by denizyuret**
This fixes #101 and #103.
- [x] Most tests pass.
- [x] Need to fix warnings.
- [x] broadcast system does not currently respect argtype signature of primitives.
- [x] functions not defined through @primitive need attention dealing with Bcasted input.
- [x] Need to test with Knet.
- [ ] #106 does not pass @gcheck.

### CODE REVIEW

**Comment by denizyuret** [comment]
@CarloLucibello I think this solves the broadcasting of user-defined functions and is ready to review.

**Comment by denizyuret** [comment]
@ekinakyurek I think this partly fixes your #106 i.e. gives a derivative but does not pass @gcheck.

**Comment by CarloLucibello** [comment]
thanks, this seems to work very well

File: src/core.notes [hunk 1]
```diff
@@ -1,12 +1,123 @@
...
+### materialize calculates the actual result using copy:
+# materialize(bc::Broadcasted) = copy(instantiate(bc))
+# materialize(x) = x
+#
+### instantiate: adds or checks the Axes component:
+# instantiate(bc::Broadcasted{S}) = Broadcasted{S}(bc.f, bc.args, combine_axes(bc.args...))
+# instantiate(x) = x
+# 
+### copy: allocates result container and fills it using copyto!
+# copy(bc::Broadcasted) = copyto!(similar(bc, etype), bc)
```
[the remark this answers is not in the document]
Comment (denizyuret) [reply src/core.notes]: Looks like the final element type: `combine_eltypes(bc.f, bc.args)` in julia/base/broadcast.jl:778

File: src/unbroadcast.jl [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    if size(x)==size(dx)
+        return dx
+    elseif isa(value(x),Number)
+        return sum(dx)
+    elseif isa(value(dx),Number)
+        return fill!(similar(value(x)),dx)
+    else
+        d = []
+        for i=1:ndims(dx)
+            size(x,i) == size(dx,i) > 1 && continue
```
[the remark this answers is not in the document]
Comment (denizyuret) [reply src/unbroadcast.jl]: dx is being reduced to x's possibly smaller size here. So size(x,i) is either 1 or the same as size(dx,i). This line says skip dimensions where x and dx have equal size greater than 1 for the sum. We will sum for dimensions where `size(dx,i)>1` and `size(x,i)==1`. If I got rid of `> 1` in this line, I guess we would also skip dimensions where `size(dx,i)==size(x,i)==1` which are currently summed. It won't matter for the result, but do you think it'll be more efficient?

**Pull request merged**